core/block_validator.go

```diff
@@ -0,0 +1,228 @@
+	// Validate the received block's bloom with the one derived from the generated receipts.
+	// For valid blocks this should always validate to true.
+	rbloom := types.CreateBloom(receipts)
+	if rbloom != header.Bloom {
+		err = fmt.Errorf("unable to replicate block's bloom=%x", rbloom)
+		return
+	}
+
+	// Tre receipt Trie's root (R = (Tr [[H1, R1], ... [Hn, R1]]))
+	receiptSha := types.DeriveSha(receipts)
+	if receiptSha != header.ReceiptHash {
+		err = fmt.Errorf("invalid receipt root hash. received=%x calculated=%x", header.ReceiptHash, receiptSha)
+		return
+	}
+
+	// Commit state objects/accounts to a database batch and calculate
+	// the state root. The database is not modified if the root
+	// doesn't match.
+	root, batch := state.CommitBatch()
+	if header.Root != root {
+		err = fmt.Errorf("invalid merkle root: header=%x computed=%x", header.Root, root)
+		return
+	}
+	return batch, nil
+}
+
+// VarifyUncles verifies the given block's uncles and applies the Ethereum
+// consensus rules to the various block headers included; it will return an
+// error if any of the included uncle headers were invalid.
+func (sm *BlockValidator) VerifyUncles(block, parent *types.Block) error {
+	// validate that there at most 2 uncles included in this block
+	if len(block.Uncles()) > 2 {
+		return ValidationError("Block can only contain maximum 2 uncles (contained %v)", len(block.Uncles()))
+	}
+
+	uncles := set.New()
+	ancestors := make(map[common.Hash]*types.Block)
+	for _, ancestor := range sm.bc.GetBlocksFromHash(block.ParentHash(), 7) {
+		ancestors[ancestor.Hash()] = ancestor
+		// Include ancestors uncles in the uncle set. Uncles must be unique.
+		for _, uncle := range ancestor.Uncles() {
+			uncles.Add(uncle.Hash())
+		}
+	}
+	ancestors[block.Hash()] = block
+	uncles.Add(block.Hash())
+
+	for i, uncle := range block.Uncles() {
+		hash := uncle.Hash()
+		if uncles.Has(hash) {
+			// Error not unique
+			return UncleError("uncle[%d](%x) not unique", i, hash[:4])
+		}
+		uncles.Add(hash)
+
+		if ancestors[hash] != nil {
+			branch := fmt.Sprintf("  O - %x\n  |\n", block.Hash())
+			for h := range ancestors {
+				branch += fmt.Sprintf("  O - %x\n  |\n", h)
+			}
+			glog.Infoln(branch)
+			return UncleError("uncle[%d](%x) is ancestor", i, hash[:4])
+		}
+
+		if ancestors[uncle.ParentHash] == nil || uncle.ParentHash == parent.Hash() {
+			return UncleError("uncle[%d](%x)'s parent is not ancestor (%x)", i, hash[:4], uncle.ParentHash[0:4])
+		}
+
+		if err := ValidateHeader(sm.Pow, uncle, ancestors[uncle.ParentHash].Header(), true, true); err != nil {
+			return ValidationError(fmt.Sprintf("uncle[%d](%x) header invalid: %v", i, hash[:4], err))
+		}
+	}
+
+	return nil
+}
+
+func (v *BlockValidator) ValidateHeader(header, parent *types.Header, checkPow bool) error {
+	// Short circuit if the header's already known or its parent missing
+	if v.bc.HasHeader(header.Hash()) {
+		return nil
+	}
+	if parent == nil {
+		return ParentError(header.ParentHash)
+	}
```

Comment:
Maybe we could swap these two checks?
 * Checking the parent's existence is "free" compared to reaching out to the database to check the current header.
 * If the parent doesn't exist, the header's invalid anyway, don't accept it even if present due to some db leftover.